### PR TITLE
Remove use of `G0` and `__target_intrinsic` in stdlib.

### LINKED ISSUE #4053: "$G0" keyword doesn't work when used in __target_switch
**Problem description**
"$G0" keyword works when used in __target_intrinsic.
But it doesn't work when used in __target_switch.

**Repo steps**
The following function exists in core.meta.slang, and works as it is.
```
__generic<T>
__target_intrinsic(cuda, "SLANG_ALIGN_OF($G0)")
__target_intrinsic(cpp, "SLANG_ALIGN_OF($G0)")
[__readNone]
[require(cpp_cuda)]
int __alignOf();
```
But when it is converted to use __target_switch, it fails on one of tests, "tests/cuda/cuda-layout.slang".
```
__generic<T>
[__readNone]
[require(cpp_cuda)]
int __alignOf()
{
    __target_switch
    {
    case cpp: __intrinsic_asm "SLANG_ALIGN_OF($G0)";
    case cuda: __intrinsic_asm "SLANG_ALIGN_OF($G0)";
    }
}
```

The same problem is observed with the following functions as well,

- __sizeOf()
- Ptr<T> TensorView::data_ptr();
- Ptr<T> TensorView::data_ptr_at();
- T TensorView::load(uint x);
- T TensorView::load(uint x, uint y);
- T TensorView::load(uint x, uint y, uint z);
- T TensorView::load(uint x, uint y, uint z, uint w);
- void TensorView::store(uint x, T val);
- void TensorView::store(uint x, uint y, T val);
- void TensorView::store(uint x, uint y, uint z, T val);
- void TensorView::store(uint x, uint y, uint z, uint w, T val);

When ran with a debug build, it fails on the following assertion,
```
const char* IntrinsicExpandContext::_emitSpecial(const char* cursor)
{
...omit...
        case 'G':
        {
            // Get the type/value at the index of the specialization of this generic

            Index argIndex = parseNat() + m_argIndexOffset;
            
            IRSpecialize* specialize = as<IRSpecialize>(m_callInst->getCallee());
            SLANG_ASSERT(specialize);
```
"specialzie" is supposed to be `kIROp_Specialize`, but it is `null_ptr` when the issue is reproduced.

### LINKED ISSUE #4093: Use SMin/UMin for SPIR-V target
Falcor unit test is failing when "min(int,int)" function uses SMin/UMin for SPIR-V target.
```
__generic<T : __BuiltinIntegerType>
[__readNone]
[require(cpp_cuda_glsl_hlsl_spirv, sm_2_0_GLSL_140)]
T min(T x, T y)
{
    __target_switch
    {
    case hlsl: __intrinsic_asm "min";
    case glsl: __intrinsic_asm "min";
    case cuda: __intrinsic_asm "$P_min($0, $1)";
    case cpp: __intrinsic_asm "$P_min($0, $1)";
#if 0 // <==== THIS NEEDS TO BE ENABLED.
    case spirv:
        if (__isSignedInt<T>())
            return spirv_asm {
                OpExtInst $$T result glsl450 SMin $x $y
            };
        else
            return spirv_asm {
                OpExtInst $$T result glsl450 UMin $x $y
            };
#endif
    default:
        return (x < y) ? x : y;
    }
}
```
When the code above is enabled, CI fails on the following Falcor test, which appears to be an application crash.
```
[----------] 12 tests from SampleGeneratorTests.cpp
[ RUN      ] SampleGeneratorTests.cpp:SampleGenerator_HeitzBelcour (D3D12)
[       OK ] SampleGeneratorTests.cpp:SampleGenerator_HeitzBelcour (D3D12) (3494 ms)
[ RUN      ] SampleGeneratorTests.cpp:SampleGenerator_HeitzBelcour (Vulkan)
[       OK ] SampleGeneratorTests.cpp:SampleGenerator_HeitzBelcour (Vulkan) (3566 ms)
[ RUN      ] SampleGeneratorTests.cpp:SampleGenerator_LeapedHalton (D3D12)
[       OK ] SampleGeneratorTests.cpp:SampleGenerator_LeapedHalton (D3D12) (1067 ms)
[ RUN      ] SampleGeneratorTests.cpp:SampleGenerator_LeapedHalton (Vulkan)
```
It can be locally reproduced only when a debug build of slang.dll is used.

### PR DESCRIPTION
**Pull request by csyonghe**
Closes #4053.
Closes #4093.

This PR extends the `__intrinsic_asm` construct to allow additional exprs as arguments, that can be referenced through the intrinsic text in the form of `"$[0]"`. For example:

```
void f<T>()
{
     __intrinsic_asm "$[0]()", T;
}
```

Will cause `f<int>()` to be emitted as `int()`.

This removes the need for `G0` in the intrinsic text, where the generic argument may no longer be accessible in the expansion context.

With this PR, we have removed all use of `__target_intrinsic` on our stdlib functions. They are still being used on types and fields, which continues to be their intended use.

### CODE REVIEW

**Review by jkwak-work — DISMISSED** [review]
Looks good to me.
The new syntax can come in handy later.

File: source/slang/hlsl.meta.slang [hunk 1]
```diff
@@ -9894,15 +9894,23 @@ vector<T,N> fmax3(vector<T,N> x, vector<T,N> y, vector<T,N> z)
...
+    case glsl:
+    case metal:
+        __intrinsic_asm "min";
+    case cuda:
+    case cpp:
+        __intrinsic_asm "$P_min($0, $1)";
+    case spirv:
+        __intrinsic_asm "OpExtInst resultType resultId glsl450 fus(FMin, UMin, SMin) _0 _1";
+    }
+}
```
Comment (jkwak-work) [inline source/slang/hlsl.meta.slang:9913]: This change may trigger an unexpected issue described in #4093
https://github.com/shader-slang/slang/issues/4093
If CI fails, try again without this change.

Comment (csyonghe) [reply source/slang/hlsl.meta.slang]: That's really weird, they should generate the same spirv.

Comment (csyonghe) [reply source/slang/hlsl.meta.slang]: This is fixed. The reason is that if there is a global `static const int x = min(1,2)`, we will end up with an `SPIRVAsm` inst in global scope and our spirv emit logic can't handle that.

The fix is to make sure we inline all global scope spirvasms into their use sites.

Comment (jkwak-work) [reply source/slang/hlsl.meta.slang]: Interesting that 4093 is also resolved...
I am not sure what the cause was but I am happy not to worry about it anymore.
Great job.

**Comment by csyonghe** [comment]
Yes, I expect we'll find more use for the syntax down the road.
Thanks for reviewing!

**Pull request merged**